### PR TITLE
Implement tensor product code:

### PR DESCRIPTION
**Pull request by Tiernan8r**
Implement the `tensor_product()` method:

Method takes in two matrices and computes the tensor product as a larger
matrix, and returns that matrix.

The tensor product of two matrices A, B is defined as:

```
A x B = [ A00 * B, A01 * B, ... A0n * B ]
        [ A10 * B, A11 * B, ... A1n * B ]
        [ ...                   ...     ]
        [ Am0 * B, An1 * B, ... Amn * B ]
```

Resulting in an `(mp)*(nq)` matrix, if A is `m*n` & B is `p*q`.

The code works by iterating over the indices of the full `mp*np` matrix,
and populating the entries. To calculate the factor of A to use, the
current indices are floor divided by the length/width of A (= n & m). 
This factor is then multiplied against the corresponding B entry, which 
is calculated by taking the modulus of the indices against it's dimensions 
(p & q).

To make reviewing the code changes easier, the following smaller changes
have been spun out to their own PRs:
* https://github.com/Tiernan8r/quantum_computing_project/pull/19
* https://github.com/Tiernan8r/quantum_computing_project/pull/20
* https://github.com/Tiernan8r/quantum_computing_project/pull/21
* https://github.com/Tiernan8r/quantum_computing_project/pull/22

### CODE REVIEW

File: mypy.ini [hunk 1]
```diff
@@ -0,0 +1,2 @@
+[mypy]
```
Comment (Tiernan8r) [inline mypy.ini:1]: Spun out to https://github.com/Tiernan8r/quantum_computing_project/pull/19

File: qcp/main.py [hunk 2]
```diff
@@ -1,4 +1,18 @@
-#!/usr/bin/env python
+#!/usr/bin/env python3
+
+# # Copyright 2022 Tiernan8r
```
Comment (Tiernan8r) [inline qcp/main.py:3]: License additions spun out to https://github.com/Tiernan8r/quantum_computing_project/pull/22

File: tox.ini [hunk 3]
```diff
@@ -18,7 +18,7 @@ commands = pytest --cache-clear --verbose --cov=qcp \
               --cov-report=xml:coverage/python/coverage.xml \
               --cov-report=term \
               --junitxml=test-reports/pytest/results.xml {posargs} \
-              {toxinidir}/qcp/tests
+              {toxinidir}/tests
```
Comment (Tiernan8r) [inline tox.ini:21]: Spun out to https://github.com/Tiernan8r/quantum_computing_project/pull/21

**Review by Tiernan8r — COMMENTED** [review]
Certain changes have been broken down into smaller PRs:

File: setup.py [hunk 4]
```diff
@@ -22,7 +22,8 @@
     name='quantum_computing_project',
     use_scm_version=True,  # tags the project version from the git tag
     setup_requires=['setuptools_scm', 'pytest-runner'],
-    packages=setuptools.find_packages(exclude=['docs', 'tests']),
+    packages=setuptools.find_packages(where="qcp", exclude=['docs', 'tests']),
+    package_dir={"": "qcp"},
```
Comment (Tiernan8r) [inline setup.py:26]: Changes were required to get the test code to discover python modules correctly.

**Review by hyoong — APPROVED** [review]

**Review by hyoong — APPROVED** [review]

**Pull request merged**